lapack/gonum/dlagv2.go

```diff
@@ -0,0 +1,209 @@
+//  [  0  a22 ]    [ -SNL  CSL ] [ a21 a22 ] [  SNR  CSR ]
+//
+//  [ b11 b12 ] := [  CSL  SNL ] [ b11 b12 ] [  CSR -SNR ]
+//  [  0  b22 ]    [ -SNL  CSL ] [  0  b22 ] [  SNR  CSR ],
+// 2) if the pencil (A,B) has a pair of complex conjugate eigenvalues, then
+//  [ a11 a12 ] := [  CSL  SNL ] [ a11 a12 ] [  CSR -SNR ]
+//  [ a21 a22 ]    [ -SNL  CSL ] [ a21 a22 ] [  SNR  CSR ]
+//
+//  [ b11  0  ] := [  CSL  SNL ] [ b11 b12 ] [  CSR -SNR ]
+//  [  0  b22 ]    [ -SNL  CSL ] [  0  b22 ] [  SNR  CSR ]
+// where b11 >= b22 > 0.
+//
+// On exit, A is overwritten by the A-part of the generalized Schur form
+// and B is overwritten by the B-part of the generalized Schur form.
+//
+// (alphar(k)+imag*alphai(k))/beta(k) are the eigenvalues of the
+// pencil (A,B), k=0,1 where imag = sqrt(-1). Note that beta(k) may
+// be zero. They are solely outputs and will be overwritten.
+//
+// Returned floats are cosines (cs) and sines (sn) of left (l) and right (r)
+// rotation matrices.
+//
+// Dlagv2 is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dlagv2(a []float64, lda int, b []float64, ldb int, alphar, alphai, beta []float64) (csl, snl, csr, snr float64) {
+	const safmin = dlamchS
+	const ulp = dlamchP
+
+	// Shorthand float64 routines.
+	var (
+		max = math.Max
+		abs = math.Abs
+	)
+	// Scale A.
+	anorm := max(abs(a[0*lda+0])+abs(a[1*lda+0]),
+		abs(a[0*lda+1])+abs(a[1*lda+1]))
+	anorm = max(anorm, safmin)
+	ascale := 1. / anorm
+	a[0*lda+0] *= ascale
+	a[0*lda+1] *= ascale
+	a[1*lda+0] *= ascale
+	a[1*lda+1] *= ascale
+
+	// Scale B.
+	bnorm := max(abs(b[0*ldb+0]),
+		abs(b[0*ldb+1])+abs(b[1*ldb+1]))
+	bnorm = max(bnorm, safmin)
+	bscale := 1. / bnorm
+	b[0*lda+0] *= bscale
+	b[0*lda+1] *= bscale
+	b[1*lda+1] *= bscale
+
+	bi := blas64.Implementation()
+	// Values returned by Dlag2 used to calculate rotation matrix.
+	var wi, wr1, scale1 float64
+	switch {
+	// Check if A can be deflated.
+	case abs(a[1*lda+0]) <= ulp:
+		// wi,snr,snl = 0,0,0 (are already zero)
+		csl, csr = 1, 1
+		a[1*lda+0] = 0
+		b[1*lda+0] = 0
+
+		// Check if B is singular.
+	case abs(b[0*lda+0]) <= ulp:
+		// snr,wi = 0,0 (are already zero)
+		csl, snl, _ = impl.Dlartg(a[0*lda+0], a[1*lda+0])
+		csr = 1
+		bi.Drot(2, a[0*lda+0:], 1, a[1*lda+0:], 1, csl, snl)
+		bi.Drot(2, b[0*ldb+0:], 1, b[1*ldb+0:], 1, csl, snl)
+		a[1*lda+0] = 0
+		b[0*ldb+0] = 0
+		b[1*ldb+0] = 0
+
+	case abs(b[1*ldb+1]) <= ulp:
+		// snl,wi = 0,0 (are already zero)
+		csr, snr, _ = impl.Dlartg(a[1*lda+1], a[1*lda+0])
+		snr *= -1
+		bi.Drot(2, a[0*lda+0:], lda, a[0*lda+1:], lda, csr, snr)
+		bi.Drot(2, b[0*ldb+0:], ldb, b[0*ldb+1:], ldb, csr, snr)
+		csl = 1
+		a[1*lda+0] = 0
+		b[1*ldb+0] = 0
+		b[1*ldb+1] = 0
+
+		// B is nonsingular, first compute the eigenvalues of (A,B).
+	default:
+		// scale1, _, wr1, _, wi = impl.Dlag2(a, lda, b, ldb)
+		csl, snl, csr, snr = impl.dlagv2aux(a, lda, b, ldb, scale1, wr1, wi)
```

Comment:
I'm not convinced that this is necessary, but I'll leave the decision up to @vladimir-ch.